src/components/structures/MessagePanel.js

```diff
@@ -382,6 +384,91 @@ module.exports = createReactClass({
             const eventId = mxEv.getId();
             const last = (mxEv === lastShownEvent);
 
+            // Wrap initial room creation events into an EventListSummary
+            // Grouping only events sent by the same user that sent the `m.room.create` and only until
+            // the first non-state event or membership event which is not regarding the sender of the `m.room.create` event
+            const shouldGroup = (ev) => {
+               if (ev.getType() === "m.room.member"
+                   && (ev.getStateKey() !== mxEv.getSender() || ev.getContent()["membership"] !== "join")) {
+                   return false;
+               }
+               if (ev.isState() && ev.getSender() === mxEv.getSender()) {
+                   return true;
+               }
+               return false;
+            };
+            if (mxEv.getType() === "m.room.create") {
+                let readMarkerInSummary = false;
+                const ts1 = mxEv.getTs();
+
+                if (this._wantsDateSeparator(prevEvent, mxEv.getDate())) {
+                    const dateSeparator = <li key={ts1+'~'}><DateSeparator key={ts1+'~'} ts={ts1} /></li>;
+                    ret.push(dateSeparator);
+                }
+
+                // If RM event is the first in the summary, append the RM after the summary
+                if (mxEv.getId() === this.props.readMarkerEventId) {
+                    readMarkerInSummary = true;
+                }
+
+                const summarisedEvents = []; // Don't add m.room.create here as we don't want it inside the summary
+                for (;i + 1 < this.props.events.length; i++) {
```

Comment:
> Not to mention that it makes the outer for loop rather long.

Agreed, but I don't see any quick wins here without rewriting the outer loop entirely

> otherwise it would break for an m.room.create after an aggregated member list because we'd have already done that check for a different event by the time we got to the second summarise loop

I can't picture this failure mode

Which check do you mean by

> that check

Reply:
Sorry - by, 'that check' I mean `if (mxEv.getType() === "m.room.create") {` - where if we're processing member events (`if (isMembershipChange(mxEv) && wantTile) {`) and find the end of them, we'll then not run the `if (mxEv.getType() === "m.room.create") {` check on the event after this member events. This happens to be OK because m.room.create events will never come after member events, but I can see someone c+ping this loop again for different event types.

Reply:
I agree this method is getting *very* long. Perhaps a good quick win would be to put both the room-create-collapse loop and the membership-collapse loop each in their own method?